### PR TITLE
Implement a single centralized master store

### PR DESCRIPTION
**Pull request by eapache**
Needs tests, otherwise ready-to-go.

@boourns @lemonmade @bouk @carsonb

### CODE REVIEW

**Comment by lemonmade** [comment]
At a high level, what is the reason to use a centralized store?

File: src/__tests__/createConnect-test.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 jest.dontMock('../dispatcher');
 jest.dontMock('../createStore');
 jest.dontMock('../createConnect');
+jest.dontMock('../masterStore');
 jest.dontMock('immutable');
 jest.dontMock('react');
```
Comment (lemonmade) [inline src/__tests__/createConnect-test.js]: This line is a pretty big indictment of Jest's mock-the-world philosophy, assuming we are doing things as well as we can.

File: src/masterStore.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    return x.every((key, value) => {
+      return isRecursivelyImmutable(key) && isRecursivelyImmutable(value);
+    });
+  }
+
+  return (type !== 'object' && type !== 'function');
+}
+
+const MasterStore: MasterStoreType = {
+  substores: new Immutable.Map(),
```
Comment (lemonmade) [inline src/masterStore.js]: No need to make this a property of the MasterStore — in fact, it's much safer to just make this a variable in scope, as that prevents anyone with access to the MasterStore from messing with the substores.

**Comment by lemonmade** [comment]
One small comment + tests = this is great :+1:

**Comment by eapache** [comment]
A centralized store serializes all state changes, which allows us to (eventually) use redux's rewind and replay features for debugging.

File: src/__tests__/createConnect-test.js (hunk 1 above)
Comment (eapache) [inline src/__tests__/createConnect-test.js]: Admittedly I didn't investigate too closely writing better mocks. I will do that once this is merged.

File: src/masterStore.js (hunk 2 above)
Comment (eapache) [inline src/masterStore.js]: There are a few of these in the dispatcher as well, I will fix them all.

File: src/__tests__/createConnect-test.js (hunk 1 above)
Comment (boourns) [inline src/__tests__/createConnect-test.js]: mock what's necessary to run the tests but don't over-mock, what's the point?

**Comment by eapache** [comment]
Comments addressed and added tests. Ready to ship on :green_apple: ?

**Comment by boourns** [comment]
yes! :ship:

**Pull request merged**